File: backend/app/services/checks.py
"""Availability checks: HTTP(S) and TCP probes with latency measurement.

Runs the configured checks (`SYSSTATUS_CHECKS`) on a fixed interval and
keeps the latest result per check. A check flips to "down" only after two
consecutive failures, so a single dropped packet never causes a false
alert.
"""

import asyncio
import logging
import time
import urllib.request
from collections import defaultdict
from datetime import datetime, timezone
from typing import Optional

from ..core.config import Settings

_FAILURES_TO_DOWN = 2
_PROBE_TIMEOUT = 5.0


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class CheckService:
    def __init__(self, settings: Settings, logger=None):
        self._checks = settings.checks
        self._interval = settings.checks_interval_seconds
        self._logger = logger or logging.getLogger("app.checks")
        self._last_run = 0.0
        self._failures: dict[str, int] = defaultdict(int)
        self._results: dict[str, dict] = {
            name: {
                "name": name,
                "target": target,
                "state": "unknown",
                "latency_ms": None,
                "checked_at": None,
                "error": None,
            }
            for name, target in self._checks
        }

    # --- Probing --------------------------------------------------------------

    async def _probe_http(self, url: str) -> tuple[float, Optional[str]]:
        def _request() -> float:
            start = time.perf_counter()
            with urllib.request.urlopen(url, timeout=_PROBE_TIMEOUT) as response:
                status = response.status
            if not (200 <= status < 400):
                raise ValueError(f"HTTP {status}")
            return (time.perf_counter() - start) * 1000

        return await asyncio.to_thread(_request)

    async def _probe_tcp(self, target: str) -> tuple[float, Optional[str]]:
        host, _, port = target.partition(":")
        if not host or not port:
            raise ValueError("destino TCP inválido (usa tcp://host:puerto)")
        start = time.perf_counter()
        reader, writer = await asyncio.wait_for(
            asyncio.open_connection(host, int(port)),
            timeout=_PROBE_TIMEOUT,
        )
        writer.close()
        try:
            await writer.wait_closed()
        except (OSError, RuntimeError):
            pass
        return (time.perf_counter() - start) * 1000

    async def _probe(self, name: str, target: str) -> dict:
        try:
            if target.startswith(("http://", "https://")):
                latency, error = await self._probe_http(target)
            elif target.startswith("tcp://"):
                latency, error = await self._probe_tcp(target.removeprefix("tcp://"))
            else:
                raise ValueError("esquema no soportado (usa http://, https:// o tcp://)")
            return {"name": name, "latency_ms": round(latency, 1), "error": error}
        except Exception as exc:  # noqa: BLE001 - any probe failure is a "down" signal
            return {"name": name, "latency_ms": None, "error": str(exc)}

    async def run_if_due(self) -> None:
        if not self._checks:
            return
        now = time.monotonic()
        if now - self._last_run < self._interval:
            return
        await self.run_once()

    async def run_once(self) -> None:
        self._last_run = time.monotonic()
        results = await asyncio.gather(
            *[self._probe(name, target) for name, target in self._checks]
        )
        for probe in results:
            name = probe["name"]
            if probe["error"] is None:
                self._failures[name] = 0
                self._results[name].update(
                    {
                        "state": "up",
                        "latency_ms": probe["latency_ms"],
                        "checked_at": _now_iso(),
                        "error": None,
                    }
                )
            else:
                self._failures[name] += 1
                down = self._failures[name] >= _FAILURES_TO_DOWN
                self._results[name].update(
                    {
                        "state": "down" if down else self._results[name]["state"],
                        "latency_ms": None,
                        "checked_at": _now_iso(),
                        "error": probe["error"],
                    }
                )
                self._logger.warning(
                    "check failed",
                    extra={"check": name, "failures": self._failures[name], "error": probe["error"]},
                )

    # --- Snapshot --------------------------------------------------------------

    def snapshot(self) -> list[dict]:
        return list(self._results.values())

    def enabled(self) -> bool:
        return bool(self._checks)

File: backend-node/src/checks/checks.ts
import * as net from 'net';
import type { Settings } from '../config/settings';

const FAILURES_TO_DOWN = 2;
const PROBE_TIMEOUT_MS = 5000;

export interface CheckResult {
  name: string;
  target: string;
  state: 'up' | 'down' | 'unknown';
  latency_ms: number | null;
  checked_at: string | null;
  error: string | null;
}

export class CheckService {
  private readonly checks: Array<{ name: string; target: string }>;
  private readonly interval: number;
  private readonly failures: Record<string, number> = {};
  private readonly results: Record<string, CheckResult> = {};
  private lastRun = 0;

  constructor(private readonly settings: Settings) {
    this.checks = settings.checks;
    this.interval = settings.checksIntervalSeconds;
    for (const { name, target } of this.checks) {
      this.results[name] = {
        name,
        target,
        state: 'unknown',
        latency_ms: null,
        checked_at: null,
        error: null,
      };
    }
  }

  private probeHttp(url: string): Promise<{ latency_ms: number } | { error: string }> {
    return new Promise((resolve) => {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), PROBE_TIMEOUT_MS);
      const started = Date.now();
      fetch(url, { signal: controller.signal, redirect: 'follow' })
        .then((response) => {
          clearTimeout(timer);
          const status = response.status;
          if (status < 200 || status >= 400) {
            resolve({ error: `HTTP ${status}` });
            return;
          }
          resolve({ latency_ms: Date.now() - started });
        })
        .catch((err) => {
          clearTimeout(timer);
          resolve({ error: err instanceof Error ? err.message : String(err) });
        });
    });
  }

  private probeTcp(target: string): Promise<{ latency_ms: number } | { error: string }> {
    return new Promise((resolve) => {
      const [host, port] = target.split(':');
      if (!host || !port) {
        resolve({ error: 'destino TCP inválido (usa tcp://host:puerto)' });
        return;
      }
      const started = Date.now();
      const socket = net.connect({
        host,
        port: Number(port),
        timeout: PROBE_TIMEOUT_MS,
      });
      socket.on('connect', () => {
        socket.destroy();
        resolve({ latency_ms: Date.now() - started });
      });
      socket.on('error', (err) => {
        socket.destroy();
        resolve({ error: err.message });
      });
      socket.on('timeout', () => {
        socket.destroy();
        resolve({ error: 'timeout' });
      });
    });
  }

  private async probe(name: string, target: string): Promise<{ name: string; latency_ms: number | null; error: string | null }> {
    try {
      if (target.startsWith('http://') || target.startsWith('https://')) {
        const result = await this.probeHttp(target);
        if ('error' in result) {
          return { name, latency_ms: null, error: result.error };
        }
        return { name, latency_ms: result.latency_ms, error: null };
      }
      if (target.startsWith('tcp://')) {
        const result = await this.probeTcp(target.replace(/^tcp:\/\//, ''));
        if ('error' in result) {
          return { name, latency_ms: null, error: result.error };
        }
        return { name, latency_ms: result.latency_ms, error: null };
      }
      return { name, latency_ms: null, error: 'esquema no soportado (usa http://, https:// o tcp://)' };
    } catch (err) {
      return { name, latency_ms: null, error: err instanceof Error ? err.message : String(err) };
    }
  }

  async runIfDue(): Promise<void> {
    if (this.checks.length === 0) return;
    const now = Date.now() / 1000;
    if (now - this.lastRun < this.interval) return;
    await this.runOnce();
  }

  async runOnce(): Promise<void> {
    this.lastRun = Date.now() / 1000;
    const results = await Promise.all(
      this.checks.map(({ name, target }) => this.probe(name, target)),
    );
    for (const probe of results) {
      const result = this.results[probe.name];
      if (!result) continue;
      if (probe.error === null) {
        this.failures[probe.name] = 0;
        result.state = 'up';
        result.latency_ms = probe.latency_ms;
        result.checked_at = new Date().toISOString();
        result.error = null;
      } else {
        this.failures[probe.name] = (this.failures[probe.name] ?? 0) + 1;
        const down = this.failures[probe.name] >= FAILURES_TO_DOWN;
        result.state = down ? 'down' : result.state;
        result.latency_ms = null;
        result.checked_at = new Date().toISOString();
        result.error = probe.error;
      }
    }
  }

  snapshot(): CheckResult[] {
    return Object.values(this.results);
  }

  enabled(): boolean {
    return this.checks.length > 0;
  }
}